File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int getmax(TreeNode* root,int& maxi){
        if(root == NULL) return 0;
        int left = getmax(root->left,maxi);
        int right = getmax(root->right,maxi);
        maxi = max(maxi,left+right);
        return 1 + max(left,right);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int maxi = INT_MIN;
         getmax(root,maxi);
         return maxi;
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode findIt(TreeNode root, TreeNode p,TreeNode q){
        if(root.val>q.val && root.val>p.val){
            return findIt(root.left,p,q);
        }
        
        if(root.val<q.val && root.val<p.val){
            return findIt(root.right,p,q);
        }
        return root;
    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return findIt(root,p,q);
    }
}
